feat: Add comment on pr when user approved request but is not member …

przechrzta-f3 opened the pull request:
Send notification when user approved PR but is not a member of any valid team.

Context: [ticket](https://github.com/form3tech/tooling-team/issues/2253)

# Proof of work
Tested on form3tech-lab
Test PR: https://github.com/form3tech-lab/gta-testing-repository/pull/1
For testing setup please read comment in [ticket](https://github.com/form3tech/tooling-team/issues/2253)

sata-form3 commented:
@przechrzta-f3 instead of using pact files, which are quite tricky to deal with in the longer run, did you consider using the integration suite which uses fakes to mimic the github interactions? See https://github.com/form3tech-oss/github-team-approver/blob/f44397ce5a65c8ef67dd63612ff7c0c1afbe26d8/internal/api/api_test.go#L75-L97 as an example.

przechrzta-f3 commented:
@sata-form3 totally agree with your  comment on pact files, tests are tricky to maintain, added test to cover comment  cleanup functionality `TestGitHubTeamApproverCleansUpOldInvalidReviewsComments`, there is also updated test `TestWhenPRReviewedByNonTeamMemberAndAuthorsArePartOfTeam` that covers basic functionality of adding comment when user is invalid.

zhenya-khvan-form3 commented on internal/api/github/client.go:
```diff
@@ -35,6 +35,9 @@ const (
 	envGitHubAppPrivateKeyPath = "GITHUB_APP_PRIVATE_KEY_PATH"
+
+	ignoredReviewersTitle = "Following reviewers have been ignored as they either contributed to or reopened the PR:\n"
+	invalidReviewersTitle = "Following reviewers have been ignored as they are not member of any valid team:\n"
```
```suggestion
	invalidReviewersTitle = "Following reviewers have been ignored as they are not a member of any valid team:\n"
```

przechrzta-f3 replied:
Updated as suggested

zhenya-khvan-form3 commented on internal/api/approval/approval.go:
```diff
@@ -164,19 +166,32 @@ func (a *Approval) ComputeApprovalStatus(ctx context.Context, pr *PR) (*Result,
+			ctx, pr.OwnerLogin, pr.RepoName, pr.Number, result.InvalidReviewers()); err != nil {
 			return nil, err
 		}
+
```
```suggestion
```

przechrzta-f3 replied:
Removed

zhenya-khvan-form3 commented on internal/api/api_test.go:
```diff
@@ -358,6 +358,31 @@ func TestGitHubTeamApproverCleansUpOldIgnoredReviewsComments(t *testing.T) {
 		ExpectedReviewRequestsMadeForFoo()
 }
 
+func TestGitHubTeamApproverCleansUpOldInvalidReviewsComments(t *testing.T) {
```
Hmm, this test and the above tests makes me wonder: what happens when GitHub Team Approver encounters ignored & invalid reviewers in the same PR?

I think it will correctly clean-up comments according to the title, but if it's easy, might be nice to add a test that covers that use-case.

przechrzta-f3 replied:
Yes, it is based on comment title

zhenya-khvan-form3 reviewed (approved):
Awesome work, and great job finding a sane way to test!

Comments are very minor and aren't blockers. :)

przechrzta-f3 commented on internal/api/stages/api_stage.go:
```diff
@@ -21,6 +21,7 @@ const (
 	httpHeaderXFinalStatus = "X-Final-Status"
 
 	ignoredReviewerMsg = "Following reviewers have been ignored as they either contributed to or reopened the PR:"
+	invalidReviewerMsg = "Following reviewers have been ignored as they are not member of any valid team:"
```
updated as suggested
